proposals/0000-type-functions.rst

```diff
@@ -0,0 +1,309 @@
+  need an object file (``.o``) to link to. This is the same limitation as we
+  have in Template Haskell, since we are reusing the underlying mechanism.
+
+* The function must be fully instantiated and applied to all inputs.
+  Furthermore, inputs and the output must be of **promotable** types. The set
+  of promotable types is implementation-defined, but it's going to include at
+  least ``Char``, ``Bool``, ``Natural``, ``String``, ``Maybe a``,
+  ``Either a b``, ``[a]``, ``(a, b)``, and so on.
+
+Even with those limitations in place, we expect the feature to be of great
+utility. Consider this excerpt from the documentation of the ``servant``
+library used to derive webservices from a declarative specification::
+
+  >>> -- GET /books/:isbn
+  >>> type MyApi = "books" :> Capture "isbn" Text :> Get '[JSON] Book
+
+You'll notice that the example is first given using the intuitive notation
+``GET /books/:isbn``, and then using the ``servant``-specific syntax ``:>``,
+``Capture``, and so on.
+
+Instead, we could imagine defining the endpoint using the intuitive notation
+only::
+
+  type MyApi = Endpoint "GET /books/:isbn" (Text -> Book) '[JSON]
+
+The only problem is that we'd need a type family ``Endpoint`` that could parse
+the intuitive specification, but type-level parsing using type families is
+difficult and slows down the compiler.
+
+However, if we could write the parser at the term level, i.e. introduce a
+helper function ``parseEndpoint :: String -> [Either String String]``, and then
+use it to define ``Endpoint``, then this approach would be viable.
+
+We could define all sorts of type-level parsers, from SQL queries to CSS color
+notation, opening up new possibilities for type-level programming.
+
+Proposed Change Specification
+-----------------------------
+
+* Extend the type syntax with qualified lowercase names ``M.fn``.
+
+* Kind checking the application of a term to zero or more type arguments,
+  ``M.fn t0 t1 ... tn``, is done as follows:
+
+  1. Check that ``M.fn`` is defined in another module that can be dynamically
+     loaded, otherwise report a type error.
+  2. Check that the types of arguments ``t0 t1 ... tn`` and the result type are
+     monomorphic and **promotable** (defined below), otherwise report a type
+     error.
+  3. Instantiate and solve constraints, e.g. ``Prelude.sum [2,3,5]`` is
+     desugared into ``Prelude.sum @Natural $dNumNatural [2,3,5]``.
+     If there's no valid instantiation or constraints cannot be solved, report
+     a type error.
+
+  At least the following types are **promotable**:
+
+  * ``Char``, ``Natural``, ``Bool``, ``()``, ``Void``
+  * ``Maybe a`` if ``a`` is promotable
+  * ``[a]`` if ``a`` is promotable
+  * ``Either a b`` if ``a`` and ``b`` are promotable
+  * ``(a, b)`` if ``a`` and ``b`` are promotable
+  * ``(a, b, c)`` if ``a``, ``b`` and ``c`` are promotable
+
+  The set of promotable types can be extended in the future and must be
+  specified in the User's Guide.
```

Comment:
Yes, I am working to support user-defined sum types. It is a difficult task but I believe it to be solved. 
Here I've listed only types supported now by my prototype.

Reply:
The prototype also supports type equality constraints.

Reply:
I would suggest the proposal should describe the ideal ultimate state of the feature, not constrained by the prototype.

Reply:
I agree about writing the proposal for the final feature. It might be good to say what has been implemented, what looks implementable, and what looks out of reach, but designing a cohesive feature regardless of implementation will lead to a better feature, I think.

Reply:
I am going to update the proposal soon, thanks. I've been very busy with the prototype.

Reply:
> I agree about writing the proposal for the final feature. It might be good to say what has been implemented, what looks implementable, and what looks out of reach, but designing a cohesive feature regardless of implementation will lead to a better feature, I think.

@goldfirere, do you suggest just writing down the final design and mentioning implementation perspectives later in the `Implementation` section?